onnxruntime/core/providers/mkldnn/nn/batch_norm.cc

```diff
@@ -0,0 +1,301 @@
+// expensive to instantiate. To address thread safety, the primitives are
+// stored in a map on thread local storage.
+template <typename T>
+class BatchNormPrimitivePool : public PrimitivePool<T> {
+ public:
+  static BatchNormPrimitive<T>* Get(const BatchNormParams& params) {
+    BatchNormPrimitive<T>* primitive = 
+      dynamic_cast<BatchNormPrimitive<T>*>(
+        BatchNormPrimitivePool<T>::GetInstance().GetPrimitive(params.ToString()));
+
+    if (primitive == nullptr) {
+      auto BatchNorm_primitive = std::make_unique<BatchNormPrimitive<T>>(params);
+      primitive = BatchNorm_primitive.get();
+      BatchNormPrimitivePool<T>::GetInstance().SetPrimitive(
+        params.ToString(), std::move(BatchNorm_primitive));
+    }
+    return primitive;
+  }
+
+ private:
+  BatchNormPrimitivePool() = default;
+  ~BatchNormPrimitivePool() = default;
+
+  static BatchNormPrimitivePool& GetInstance() {
+    static BatchNormPrimitivePool pool;
+    return pool;
+  }
+};
+} // namespace
+
+template <typename T>
+Status BatchNorm<T>::Compute(OpKernelContext* context) const {
+  const Tensor* X = context->Input<Tensor>(0);
+  
+  int num_dimensions = static_cast<int>(X->Shape().NumDimensions());
+  if (num_dimensions == 3) {
+    // Fall back CPU implementation
+    return onnxruntime::BatchNorm<T>::Compute(context);
+  }
+
+  const T* src_data = X->template Data<T>();
+
+  const Tensor* scale = context->Input<Tensor>(1);
+  const T* scale_data = scale->template Data<T>();
+
+  const Tensor* B = context->Input<Tensor>(2);
+  const T* b_data = B->template Data<T>();
+
+  const Tensor* mean = context->Input<Tensor>(3);
+  const T* mean_data = mean->template Data<T>();
+
+  const Tensor* var = context->Input<Tensor>(4);
+  const T* var_data = var->template Data<T>();
+
+  Tensor* Y = context->Output(0, X->Shape());
+  T* dst_data = Y->template MutableData<T>();
+
+  ONNXRUNTIME_RETURN_IF_ERROR(
+    BatchNormHelper::ValidateInputs(X, scale, B, mean, var));
+
+  mkldnn::memory::dims src_dims_mkl(
+    X->Shape().GetDims().begin(), X->Shape().GetDims().end());
+  mkldnn::memory::dims scale_dims_mkl(
+    scale->Shape().GetDims().begin(), scale->Shape().GetDims().end());
+  mkldnn::memory::dims b_dims_mkl(
+    B->Shape().GetDims().begin(), B->Shape().GetDims().end());
+  mkldnn::memory::dims mean_dims_mkl(
+    mean->Shape().GetDims().begin(), mean->Shape().GetDims().end());
+  mkldnn::memory::dims var_dims_mkl(
+    var->Shape().GetDims().begin(), var->Shape().GetDims().end());
+
+  mkldnn::memory::dims dst_dims_mkl(
+    Y->Shape().GetDims().begin(), Y->Shape().GetDims().end());
+
+  try {
+    BatchNormParams batchNorm_params(src_dims_mkl, scale_dims_mkl, 
+      b_dims_mkl, mean_dims_mkl, var_dims_mkl, dst_dims_mkl, 
+      onnxruntime::BatchNorm<T>::epsilon_, num_dimensions);
+    BatchNormPrimitive<T>* batchNorm_primitive = 
+      BatchNormPrimitivePool<T>::Get(batchNorm_params);
+    batchNorm_primitive->Compute(src_data, scale_data, b_data, 
```

Comment:
Is nullptr check needed here?

Reply:
We are not checking for nullptr check in other primitives.  `BatchNormPrimitivePool<T>::Get `checks for batchNorm_primitive in the pool. If it is null, it will create new one and return.

I will add `ONNXRUNTIME_RETURN_IF_ERROR`. and also to Sum primitive.

Reply:
Added check.

`ONNXRUNTIME_RETURN_IF_NOT(batchNorm_primitive != nullptr);`